functional_tests/steps/bundles.py

```diff
@@ -121,3 +128,268 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+                context.bundles.append(bundle)
+
+
+def json_str_to_dict(bundle_details):
+    try:
+        return json.loads(bundle_details)
+    except ValueError as e:
+        print(e)
+        return False
+
+
+# Bundles UI Triggers
+@when("the {user_role} logs in")
+def log_in_user_by_role(context: Context, user_role: str) -> None:
+    context.page.goto(f"{context.base_url}/admin/login/")
+    context.page.get_by_placeholder("Enter your username").fill(context.users[user_role]["username"])
+    context.page.get_by_placeholder("Enter password").fill(context.users[user_role]["password"])
+    context.page.get_by_role("button", name="Sign in").click()
+
+
+# Bundles UI Consequences
+@then("the user can create a bundle")
+def add_bundle_details(context: Context) -> None:
+    bundle_name = "Bundle UI Test 1"
+    context.page.get_by_role("link", name="Add bundle").click()
+    context.page.get_by_role("textbox", name="Name*").click()
+    context.page.get_by_role("textbox", name="Name*").fill(bundle_name)
+    context.page.get_by_role("button", name="Save as draft").click()
+    context.page.get_by_role("link", name="Dashboard").click()
+
+
+@then("the user can see the created bundle")
+def can_see_bundle(context: Context) -> None:
+    bundle_name = "Bundle UI Test 1"
+    expect(context.page.locator("#latest-bundles-content")).to_contain_text("Add bundle")
+    expect(context.page.get_by_text("Latest active bundles")).to_be_visible()
+    expect(context.page.get_by_role("link", name=bundle_name)).to_be_visible()
+
+
+@then("the user cannot create a bundle")
+def cannot_add_bundles(context: Context) -> None:
+    expect(context.page.get_by_role("link", name="Add bundle")).not_to_be_visible()
+    expect(context.page.get_by_text("Bundles ready for preview")).to_be_visible()
+    expect(context.page.get_by_text("There are currently no")).to_be_visible()
+
+
+@step("the user can find the bundle")
+def can_find_bundle(context: Context) -> None:
+    expect(context.page.locator("#latest-bundles-heading")).to_contain_text("Latest active bundles")
+    expect(context.page.locator("#latest-bundles-content")).to_contain_text(context.bundles[0].name)
+
+
+@step("the user can edit the bundle")
+def can_edit_bundle(context: Context) -> None:
+    context.page.get_by_role("row", name=context.bundles[0].name + " Actions").get_by_label("Actions").click()
+    context.page.get_by_role("link", name="Edit", exact=True).click()
+
+    add_release_calendar_in_edit(context)
+    context.page.get_by_role("button", name="Save").click()
+    add_preview_team_in_edit(context)
+    context.page.get_by_role("button", name="Save").click()
+    add_release_calendar_in_edit(context)
+    context.page.get_by_role("button", name="Save").click()
+
+    context.page.get_by_role("button", name="More actions").click()
+    context.page.get_by_role("button", name="Save to preview").click()
+
+
+def add_preview_team_in_edit(context: Context) -> None:
```

Comment:
I think it might be worth moving these helper functions to a separate file in `step_helpers` just to follow our existing convention.